File: conformance/ts/models/op.mjs
// op-encoding / op-decode-negative / op-signature handlers per
// doc/KERNEL.md §4/§9.
//
// op-encoding: rebuild the §4.1 envelope map from the vector's logical
// description, encode with the independent CBOR codec, require byte
// agreement with canonical_hex plus decode→re-encode identity, and derive
// OpId = BLAKE3(domain ‖ bytes) with the independent JS BLAKE3 (I-6).
//
// op-signature: verify the Rust-emitted Ed25519 golden signature over the
// §4.4 sig-preimage using node:crypto, plus tamper negatives.

import { createPublicKey, verify as edVerify } from 'node:crypto';
import { Buffer } from 'node:buffer';

import { encode, decode, bytesToHex, hexToBytes } from './cbor.mjs';
import { blake3 } from './blake3.mjs';

const DOMAIN_OP_ID = new TextEncoder().encode('zerodb-op-id-v1');
const DOMAIN_OP_SIG = new TextEncoder().encode('zerodb-op-sig-v1');
// DER SPKI prefix for a raw Ed25519 public key.
const SPKI_PREFIX = hexToBytes('302a300506032b6570032100');

function opId(canonicalBytes) {
  const preimage = new Uint8Array(DOMAIN_OP_ID.length + canonicalBytes.length);
  preimage.set(DOMAIN_OP_ID, 0);
  preimage.set(canonicalBytes, DOMAIN_OP_ID.length);
  return blake3(preimage);
}

function verifySignature(pub32, message, sig64) {
  const key = createPublicKey({
    key: Buffer.concat([Buffer.from(SPKI_PREFIX), Buffer.from(pub32)]),
    format: 'der',
    type: 'spki',
  });
  return edVerify(null, Buffer.from(message), key, Buffer.from(sig64));
}

/** HEX body keys that encode as CBOR bytes (same list as Rust `json_to_cbor_body`). */
export const HEX_BODY_KEYS = new Set([
  'node',
  'edge',
  'src',
  'dst',
  'founder',
  'salt',
  'subject',
  'ds_bind',
  'grant',
  'encrypted',
  'key_id',
  'recipient',
  'eph_pk',
  'nonce',
  'wrapped',
  'device',
  'principal',
  'root_pk',
  'cert_sig',
  'revoke_of',
  'schema',
  'ir',
  'prev',
]);

/** JSON WireOp body → tagged CBOR matching `json_to_cbor_body`. */
export function jsonToBodyTagged(val, key) {
  if (val === null || val === undefined) return { t: 'null' };
  if (typeof val === 'boolean') return { t: 'bool', v: val };
  if (typeof val === 'number') {
    if (!Number.isInteger(val) || val < 0) throw new Error(`non-uint ${val}`);
    return { t: 'uint', v: val };
  }
  if (typeof val === 'string') {
    if (key && HEX_BODY_KEYS.has(key)) {
      if (!/^[0-9a-f]*$/i.test(val) || val.length % 2 !== 0) {
        throw new Error(`hex body field ${key} must be even hex`);
      }
      return { t: 'bytes', hex: val.toLowerCase() };
    }
    return { t: 'text', v: val };
  }
  if (Array.isArray(val)) {
    return { t: 'array', v: val.map((item) => jsonToBodyTagged(item)) };
  }
  if (typeof val === 'object') {
    const v = {};
    for (const [k, item] of Object.entries(val)) v[k] = jsonToBodyTagged(item, k);
    return { t: 'map', v };
  }
  throw new Error(`unsupported body ${typeof val}`);
}

export function envelopeToTagged(op) {
  const body = op.body && op.body.t ? op.body : jsonToBodyTagged(op.body);
  return {
    t: 'map',
    v: {
      v: { t: 'uint', v: op.v },
      ds: { t: 'bytes', hex: op.ds },
      ep: { t: 'uint', v: op.ep },
      author: { t: 'bytes', hex: op.author },
      ts: {
        t: 'map',
        v: {
          l: { t: 'uint', v: op.ts.l },
          p: { t: 'uint', v: op.ts.p },
        },
      },
      deps: { t: 'array', v: op.deps.map((d) => ({ t: 'bytes', hex: d })) },
      grp: op.grp === null || op.grp === undefined ? { t: 'null' } : { t: 'bytes', hex: op.grp },
      kind: { t: 'uint', v: op.kind },
      body,
    },
  };
}

export function runOpEncodingVector(vector) {
  const bytes = encode(envelopeToTagged(vector.op));
  const hex = bytesToHex(bytes);
  if (hex !== vector.canonical_hex) {
    throw new Error(
      `canonical bytes mismatch:\n  expected ${vector.canonical_hex}\n  got      ${hex}`
    );
  }
  const reencoded = bytesToHex(encode(decode(bytes)));
  if (reencoded !== hex) {
    throw new Error('decode -> re-encode is not byte-identical');
  }
  const idHex = bytesToHex(opId(bytes));
  if (idHex !== vector.op_id_hex) {
    throw new Error(`op id mismatch:\n  expected ${vector.op_id_hex}\n  got      ${idHex}`);
  }
}

export function runOpSignatureVector(vector) {
  const canonical = encode(envelopeToTagged(vector.op));
  const id = opId(canonical);
  const idHex = bytesToHex(id);
  if (idHex !== vector.op_id_hex) {
    throw new Error(`op id mismatch: expected ${vector.op_id_hex}, got ${idHex}`);
  }

  const preimage = new Uint8Array(DOMAIN_OP_SIG.length + id.length);
  preimage.set(DOMAIN_OP_SIG, 0);
  preimage.set(id, DOMAIN_OP_SIG.length);

  const pub = hexToBytes(vector.pub_hex);
  const sig = hexToBytes(vector.sig_hex);
  if (!verifySignature(pub, preimage, sig)) {
    throw new Error('golden signature failed to verify');
  }

  // Tamper negatives: a flipped signature byte and a flipped preimage byte
  // must both fail verification (I-8).
  const badSig = Uint8Array.from(sig);
  badSig[0] ^= 1;
  if (verifySignature(pub, preimage, badSig)) {
    throw new Error('tampered signature verified');
  }
  const badPreimage = Uint8Array.from(preimage);
  badPreimage[DOMAIN_OP_SIG.length] ^= 1;
  if (verifySignature(pub, badPreimage, sig)) {
    throw new Error('signature verified over a tampered op id');
  }
}

export function runOpDecodeNegativeVector(vector) {
  let error = null;
  try {
    decode(hexToBytes(vector.raw_hex));
  } catch (e) {
    error = e.message;
  }
  if (error === null) {
    throw new Error(`expected ${vector.expect_error}, but decode succeeded`);
  }
  if (error !== vector.expect_error) {
    throw new Error(`expected ${vector.expect_error}, got ${error}`);
  }
}

export function computeOpId(canonicalBytes) {
  return opId(canonicalBytes);
}

export function sigPreimage(id) {
  const preimage = new Uint8Array(DOMAIN_OP_SIG.length + id.length);
  preimage.set(DOMAIN_OP_SIG, 0);
  preimage.set(id, DOMAIN_OP_SIG.length);
  return preimage;
}

export { verifySignature };

File: zerodb-core/src/op.rs
//! Operation envelope per doc/KERNEL.md §4.1 and `OpId` derivation per §4.4.
//!
//! The envelope encodes as one canonical CBOR map (keys "v", "ds", "ep",
//! "author", "ts", "deps", "grp", "kind", "body" [+ "id", "sig" in the full
//! form]). The id-preimage is the map WITHOUT "id"/"sig", prefixed by the
//! domain-separation string; `OpId = BLAKE3(id-preimage)`. Signatures
//! (Ed25519 over the domain-separated OpId) land with the crypto layer.

use crate::cbor::{Cbor, CborError, encode};

/// Domain-separation prefix for the id-preimage (registry `domain_separation.op_id`).
pub const DOMAIN_OP_ID: &[u8] = b"zerodb-op-id-v1";

/// HLC timestamp as carried in the envelope ("ts": {"l": .., "p": ..});
/// the peer component is implied by "author".
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct OpTs {
    pub physical_ms: u64,
    pub logical: u16,
}

/// KERNEL §4.1 envelope, minus "id"/"sig" (which are derived/attached).
#[derive(Debug, Clone)]
pub struct OpEnvelope {
    pub v: u64,
    pub ds: [u8; 32],
    pub ep: u64,
    pub author: [u8; 32],
    pub ts: OpTs,
    pub deps: Vec<[u8; 32]>,
    pub grp: Option<[u8; 16]>,
    pub kind: u64,
    pub body: Cbor,
}

impl OpEnvelope {
    fn as_cbor(&self) -> Cbor {
        Cbor::Map(vec![
            ("v".into(), Cbor::Uint(self.v)),
            ("ds".into(), Cbor::Bytes(self.ds.to_vec())),
            ("ep".into(), Cbor::Uint(self.ep)),
            ("author".into(), Cbor::Bytes(self.author.to_vec())),
            (
                "ts".into(),
                Cbor::Map(vec![
                    ("l".into(), Cbor::Uint(self.ts.logical as u64)),
                    ("p".into(), Cbor::Uint(self.ts.physical_ms)),
                ]),
            ),
            (
                "deps".into(),
                Cbor::Array(self.deps.iter().map(|d| Cbor::Bytes(d.to_vec())).collect()),
            ),
            (
                "grp".into(),
                match &self.grp {
                    Some(g) => Cbor::Bytes(g.to_vec()),
                    None => Cbor::Null,
                },
            ),
            ("kind".into(), Cbor::Uint(self.kind)),
            ("body".into(), self.body.clone()),
        ])
    }

    /// Canonical bytes of the envelope map without "id"/"sig" (KERNEL §4.4).
    pub fn id_preimage_body(&self) -> Result<Vec<u8>, CborError> {
        encode(&self.as_cbor())
    }

    /// `OpId = BLAKE3(domain ‖ canonical bytes)`.
    pub fn op_id(&self) -> Result<[u8; 32], CborError> {
        let mut hasher = blake3::Hasher::new();
        hasher.update(DOMAIN_OP_ID);
        hasher.update(&self.id_preimage_body()?);
        Ok(*hasher.finalize().as_bytes())
    }
}

/// Experimental WireOp JSON body → CBOR (same rules as LocalStore).
/// Hex strings on `node` / `edge` / `src` / `dst` become bytes.
/// AUTH control bodies also encode `founder` / `salt` / `subject` / `ds_bind` /
/// `grant` as bytes, and numeric arrays (grant `scopes`) as uints.
/// Group-key wraps (`kr = 2`) and KERNEL §7 envelopes recurse into maps.
pub fn json_to_cbor_body(v: &serde_json::Value) -> Result<Cbor, String> {
    match v {
        serde_json::Value::Object(_) => json_to_cbor(v, None),
        _ => Ok(Cbor::Map(vec![])),
    }
}

fn json_to_cbor(val: &serde_json::Value, key: Option<&str>) -> Result<Cbor, String> {
    match val {
        serde_json::Value::String(s) => {
            if key.is_some_and(is_hex_body_key) {
                let b = hex::decode(s).map_err(|e| e.to_string())?;
                Ok(Cbor::Bytes(b))
            } else {
                Ok(Cbor::Text(s.clone()))
            }
        }
        serde_json::Value::Number(n) => {
            if let Some(u) = n.as_u64() {
                Ok(Cbor::Uint(u))
            } else {
                Ok(Cbor::Text(n.to_string()))
            }
        }
        serde_json::Value::Bool(b) => Ok(Cbor::Bool(*b)),
        serde_json::Value::Null => Ok(Cbor::Null),
        serde_json::Value::Array(a) => {
            let mut items = Vec::new();
            for x in a {
                items.push(json_to_cbor(x, None)?);
            }
            Ok(Cbor::Array(items))
        }
        serde_json::Value::Object(o) => {
            let mut entries = Vec::new();
            for (k, v) in o {
                entries.push((k.clone(), json_to_cbor(v, Some(k))?));
            }
            Ok(Cbor::Map(entries))
        }
    }
}

fn is_hex_body_key(k: &str) -> bool {
    matches!(
        k,
        "node"
            | "edge"
            | "src"
            | "dst"
            | "founder"
            | "salt"
            | "subject"
            | "ds_bind"
            | "grant"
            | "encrypted"
            | "key_id"
            | "recipient"
            | "eph_pk"
            | "nonce"
            | "wrapped"
            | "device"
            | "principal"
            | "root_pk"
            | "cert_sig"
            | "revoke_of"
    )
}

#[cfg(test)]
mod tests {
    use super::*;

    fn sample() -> OpEnvelope {
        OpEnvelope {
            v: 1,
            ds: [0x11; 32],
            ep: 0,
            author: [0x22; 32],
            ts: OpTs {
                physical_ms: 1_000_000,
                logical: 3,
            },
            deps: vec![[0x33; 32]],
            grp: None,
            kind: 1,
            body: Cbor::Map(vec![
                ("label".into(), Cbor::Text("User".into())),
                ("node".into(), Cbor::Bytes(vec![0x44; 16])),
            ]),
        }
    }

    #[test]
    fn op_id_is_stable_and_field_sensitive() {
        let a = sample();
        let id1 = a.op_id().unwrap();
        assert_eq!(id1, a.op_id().unwrap());

        let mut b = sample();
        b.ep = 1;
        assert_ne!(id1, b.op_id().unwrap(), "epoch must be in the preimage");

        let mut c = sample();
        c.ds[0] ^= 1;
        assert_ne!(id1, c.op_id().unwrap(), "datastore must be in the preimage");
    }

    #[test]
    fn preimage_round_trips_canonically() {
        let bytes = sample().id_preimage_body().unwrap();
        let decoded = crate::cbor::decode(&bytes).unwrap();
        assert_eq!(crate::cbor::encode(&decoded).unwrap(), bytes);
    }
}
